File: tools/content.py
#!/usr/bin/env python3
"""Validate school content and regenerate docs/schools.md.

Usage: python3 tools/content.py [--docs]
"""
import json, pathlib, re, sys

ROOT = pathlib.Path(__file__).resolve().parent.parent
SCHOOLS = ROOT / "app/src/main/assets/content/schools"
EMOTIONS = ("anxiety fear sadness grief loneliness anger shame guilt envy emptiness confusion fatigue boredom "
            "joy love gratitude hope nostalgia craving addiction hangover euphoria crash panic dread fomo sleepless "
            "betrayal jealousy resentment heartbreak impostor self_hatred helplessness procrastination uprooted "
            "overload unreality awe relief calm inspiration").split()
FAMILIES = {
    "classic": "Античность и Восток", "early_modern": "XVI–XVIII века", "nineteenth": "XIX век",
    "phenomenology_existential": "Феноменология и экзистенциализм",
    "psychoanalysis_critical": "Психоанализ и критическая теория",
    "analytic": "Аналитическая философия и прагматизм",
    "poststructuralism": "Постструктурализм", "affect_theory": "Теория аффекта и наследники",
    "contemporary": "XXI век",
}
REQUIRED = ["id", "title", "tradition", "family", "period", "sortYear", "thinkers", "keyWorks", "summary", "onEmotions", "entries"]


def load():
    schools, errors = [], []
    for f in sorted(SCHOOLS.glob("*.json")):
        try:
            d = json.loads(f.read_text(encoding="utf-8"))
        except Exception as e:
            errors.append(f"{f.name}: invalid JSON: {e}")
            continue
        for k in REQUIRED:
            if k not in d:
                errors.append(f"{f.name}: missing {k}")
        if d.get("id") != f.stem:
            errors.append(f"{f.name}: id {d.get('id')!r} != file name")
        if d.get("family") not in FAMILIES:
            errors.append(f"{f.name}: unknown family {d.get('family')!r}")
        entries = d.get("entries", {})
        missing = [e for e in EMOTIONS if e not in entries]
        extra = [e for e in entries if e not in EMOTIONS]
        if missing or extra:
            errors.append(f"{f.name}: missing {missing} extra {extra}")
        for eid, e in entries.items():
            if not e.get("text", "").strip() or not e.get("practice", "").strip():
                errors.append(f"{f.name}/{eid}: empty text or practice")
            q = e.get("quote")
            if q is not None and not (isinstance(q, dict) and q.get("text") and q.get("author")):
                errors.append(f"{f.name}/{eid}: malformed quote")
        schools.append(d)
    return schools, errors


def words(s):
    return len(re.findall(r"\w+", s))


def docs(schools):
    order = list(FAMILIES)
    schools = sorted(schools, key=lambda s: (order.index(s["family"]), s["sortYear"]))
    total = sum(len(s["entries"]) for s in schools)
    quotes = sum(1 for s in schools for e in s["entries"].values() if e.get("quote"))
    out = ["# Школы и мыслители", "",
           f"База приложения «Внутри»: {len(schools)} школ × {len(EMOTIONS)} чувств = {total} текстов, {quotes} цитат.",
           "Файл сгенерирован `tools/content.py --docs` из `app/src/main/assets/content/schools/`.", ""]
    for fam in order:
        group = [s for s in schools if s["family"] == fam]
        if not group:
            continue
        out += [f"## {FAMILIES[fam]}", ""]
        for s in group:
            out += [f"### {s['title']} — {s['tradition']}", "",
                    f"*{s['period']}* · {', '.join(s['thinkers'])}", "",
                    f"**Ключевые работы:** {'; '.join(s['keyWorks'])}", "",
                    s["summary"], "", f"**О чувствах.** {s['onEmotions']}", ""]
    (ROOT / "docs").mkdir(exist_ok=True)
    (ROOT / "docs/schools.md").write_text("\n".join(out), encoding="utf-8")


if __name__ == "__main__":
    schools, errors = load()
    for s in schools:
        ws = [words(e["text"]) for e in s["entries"].values()]
        q = sum(1 for e in s["entries"].values() if e.get("quote"))
        print(f"{s['id']:<20} {s['family']:<26} entries={len(s['entries']):>2} words={min(ws)}-{max(ws)} quotes={q}")
    print(f"\n{len(schools)} schools, {len(errors)} errors")
    for e in errors:
        print("ERROR", e)
    if "--docs" in sys.argv and not errors:
        docs(schools)
        print("docs/schools.md written")
    sys.exit(1 if errors else 0)

File: app/src/main/java/com/ilyamalshv/vnutri/data/Content.kt
package com.ilyamalshv.vnutri.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

data class Emotion(val id: String, val name: String, val light: Boolean)

object Emotions {
    val all = listOf(
        Emotion("anxiety", "Тревога", false),
        Emotion("fear", "Страх", false),
        Emotion("sadness", "Грусть", false),
        Emotion("grief", "Горе, утрата", false),
        Emotion("loneliness", "Одиночество", false),
        Emotion("anger", "Злость", false),
        Emotion("shame", "Стыд", false),
        Emotion("guilt", "Вина", false),
        Emotion("envy", "Зависть", false),
        Emotion("emptiness", "Пустота, апатия", false),
        Emotion("confusion", "Растерянность", false),
        Emotion("fatigue", "Усталость, выгорание", false),
        Emotion("boredom", "Скука", false),
        Emotion("nostalgia", "Ностальгия", true),
        Emotion("joy", "Радость", true),
        Emotion("love", "Любовь, нежность", true),
        Emotion("gratitude", "Благодарность", true),
        Emotion("hope", "Надежда", true),
    )
    private val byId = all.associateBy { it.id }

    fun name(id: String): String = byId[id]?.name ?: id
}

data class Family(val id: String, val name: String)

object Families {
    val all = listOf(
        Family("classic", "Античность и Восток"),
        Family("early_modern", "XVII–XVIII века"),
        Family("nineteenth", "XIX век"),
        Family("phenomenology_existential", "Феноменология и экзистенциализм"),
        Family("psychoanalysis_critical", "Психоанализ и критическая теория"),
        Family("poststructuralism", "Постструктурализм"),
        Family("affect_theory", "Теория аффекта и наследники"),
        Family("contemporary", "XXI век"),
    )

    fun order(id: String): Int = all.indexOfFirst { it.id == id }.let { if (it < 0) all.size else it }
    fun name(id: String): String = all.firstOrNull { it.id == id }?.name ?: id
}

data class Quote(val text: String, val author: String, val source: String)

data class Lens(val text: String, val practice: String, val quote: Quote?)

data class School(
    val id: String,
    val title: String,
    val tradition: String,
    val family: String,
    val period: String,
    val sortYear: Int,
    val thinkers: List<String>,
    val keyWorks: List<String>,
    val summary: String,
    val onEmotions: String,
    val lenses: Map<String, Lens>,
)

class Library(val schools: List<School>) {
    private val byId = schools.associateBy { it.id }

    fun school(id: String): School? = byId[id]

    fun lensesFor(emotionId: String): List<Pair<School, Lens>> =
        schools.mapNotNull { s -> s.lenses[emotionId]?.let { s to it } }

    companion object {
        private const val DIR = "content/schools"

        fun load(context: Context): Library {
            val files = context.assets.list(DIR).orEmpty().filter { it.endsWith(".json") }
            val schools = files.mapNotNull { name ->
                runCatching {
                    val raw = context.assets.open("$DIR/$name").bufferedReader().use { it.readText() }
                    parse(JSONObject(raw))
                }.getOrNull()
            }
            return Library(schools.sortedWith(compareBy({ Families.order(it.family) }, { it.sortYear })))
        }

        private fun parse(o: JSONObject): School {
            val entries = o.getJSONObject("entries")
            val lenses = mutableMapOf<String, Lens>()
            entries.keys().forEach { key ->
                val e = entries.getJSONObject(key)
                val quote = e.optJSONObject("quote")?.let { q ->
                    Quote(q.optString("text"), q.optString("author"), q.optString("source"))
                }?.takeIf { it.text.isNotBlank() }
                lenses[key] = Lens(e.getString("text"), e.optString("practice"), quote)
            }
            return School(
                id = o.getString("id"),
                title = o.getString("title"),
                tradition = o.optString("tradition"),
                family = o.optString("family"),
                period = o.optString("period"),
                sortYear = o.optInt("sortYear"),
                thinkers = o.optJSONArray("thinkers").strings(),
                keyWorks = o.optJSONArray("keyWorks").strings(),
                summary = o.optString("summary"),
                onEmotions = o.optString("onEmotions"),
                lenses = lenses,
            )
        }

        private fun JSONArray?.strings(): List<String> =
            if (this == null) emptyList() else List(length()) { getString(it) }
    }
}
